// @ngx-pwa/local-storage/esm5/lib/databases/memory-database.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { of, from } from 'rxjs';
import * as i0 from "@angular/core";
var MemoryDatabase = /** @class */ (function () {
    function MemoryDatabase() {
        /**
         * Memory storage
         */
        this.memoryStorage = new Map();
    }
    Object.defineProperty(MemoryDatabase.prototype, "size", {
        /**
         * Number of items in memory
         */
        get: function () {
            /* Wrap in a RxJS `Observable` to be consistent with other storages */
            return of(this.memoryStorage.size);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Gets an item value in memory
     * @param key The item's key
     * @returns The item's value if the key exists, `undefined` otherwise, wrapped in a RxJS `Observable`
     */
    MemoryDatabase.prototype.get = function (key) {
        var rawData = this.memoryStorage.get(key);
        /* Wrap in a RxJS `Observable` to be consistent with other storages */
        return of(rawData);
    };
    /**
     * Sets an item in memory
     * @param key The item's key
     * @param data The item's value
     * @returns A RxJS `Observable` to wait the end of the operation
     */
    MemoryDatabase.prototype.set = function (key, data) {
        this.memoryStorage.set(key, data);
        /* Wrap in a RxJS `Observable` to be consistent with other storages */
        return of(undefined);
    };
    /**
     * Deletes an item in memory
     * @param key The item's key
     * @returns A RxJS `Observable` to wait the end of the operation
     */
    MemoryDatabase.prototype.delete = function (key) {
        this.memoryStorage.delete(key);
        /* Wrap in a RxJS `Observable` to be consistent with other storages */
        return of(undefined);
    };
    /**
     * Deletes all items in memory
     * @returns A RxJS `Observable` to wait the end of the operation
     */
    MemoryDatabase.prototype.clear = function () {
        this.memoryStorage.clear();
        /* Wrap in a RxJS `Observable` to be consistent with other storages */
        return of(undefined);
    };
    /**
     * Get all keys in memory
     * @returns A RxJS `Observable` iterating on keys
     */
    MemoryDatabase.prototype.keys = function () {
        /* Create an `Observable` from keys */
        return from(this.memoryStorage.keys());
    };
    /**
     * Check if a key exists in memory
     * @param key Key name
     * @returns a RxJS `Observable` telling if the key exists or not
     */
    MemoryDatabase.prototype.has = function (key) {
        /* Wrap in a RxJS `Observable` to be consistent with other storages */
        return of(this.memoryStorage.has(key));
    };
    MemoryDatabase.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MemoryDatabase_Factory() { return new MemoryDatabase(); }, token: MemoryDatabase, providedIn: "root" });
    MemoryDatabase = tslib_1.__decorate([
        Injectable({
            providedIn: 'root'
        })
    ], MemoryDatabase);
    return MemoryDatabase;
}());
export { MemoryDatabase };
